signer/signer.go

```diff
@@ -0,0 +1,257 @@
+	RSASSA_PSS_SHA_512 SignatureAlgorithm = "RSASSA_PSS_SHA_512"
+	ECDSA_SHA_256      SignatureAlgorithm = "ECDSA_SHA_256"
+	ECDSA_SHA_384      SignatureAlgorithm = "ECDSA_SHA_384"
+	ECDSA_SHA_512      SignatureAlgorithm = "ECDSA_SHA_512"
+)
+
+// SignerInfo represents a parsed signature envelope that is agnostic to signature envelope format.
+type SignerInfo struct {
+	Payload            []byte
+	PayloadContentType string
+	SignedAttributes   SignedAttributes
+	UnsignedAttributes UnsignedAttributes
+	SignatureAlgorithm SignatureAlgorithm
+	CertificateChain   []*x509.Certificate
+	Signature          []byte
+	TimestampSignature []byte
+}
+
+// SignedAttributes represents signed metadata in the Signature envelope
+type SignedAttributes struct {
+	SigningTime        time.Time
+	Expiry             time.Time
+	ExtendedAttributes []Attribute
+}
+
+// UnsignedAttributes represents unsigned metadata in the Signature envelope
+type UnsignedAttributes struct {
+	SigningAgent string
+}
+
+// SignRequest is used to generate Signature.
+type SignRequest struct {
+	Payload             []byte
+	PayloadContentType  string
+	CertificateChain    []*x509.Certificate
+	SignatureProvider   SignatureProvider
+	SigningTime         time.Time
+	Expiry              time.Time
+	ExtendedSignedAttrs []Attribute
+	SigningAgent        string
+}
+
+type Attribute struct {
+	Key      string
+	Critical bool
+	Value    interface{}
+}
+
+// SignatureProvider is used to sign bytes generated after creating Signature envelope.
+type SignatureProvider interface {
+	Sign([]byte) ([]byte, error)
+}
+
+type SignatureEnvelope struct {
+	rawSignatureEnvelope []byte
+	internalEnvelope     internalSignatureEnvelope
+}
+
+// Contains a set of common methods that every Signature envelope format must implement.
+type internalSignatureEnvelope interface {
+	// validateIntegrity validates the integrity of given Signature envelope.
+	validateIntegrity() error
+	// getSignerInfo returns the information stored in the Signature envelope and doesn't perform integrity verification.
+	getSignerInfo() (*SignerInfo, error)
+	// signPayload created Signature envelope.
+	signPayload(SignRequest) ([]byte, error)
+}
+
+// Verify performs integrity validation and validates that cert-chain stored in Signature leads to a given set of the trusted root.
+// Returns the certificate that is used for establishing trust i.e. authenticity of the signature, SignerInfo object
+// containing the information about the signature.
+func (s *SignatureEnvelope) Verify(trustedCerts []*x509.Certificate) (*x509.Certificate, *SignerInfo, error) {
+	if len(s.rawSignatureEnvelope) == 0 {
+		return nil, nil, SignatureNotFoundError{}
+	}
+
+	if len(trustedCerts) == 0 {
+		return nil, nil, MalformedArgumentError{param: "trustedCerts"}
+	}
+
+	integrityError := s.internalEnvelope.validateIntegrity()
+	if integrityError != nil {
+		return nil, nil, integrityError
+	}
+
+	singerInfo, singerInfoErr := s.GetSignerInfo()
```

Comment:
If we end up not returning the `SignerInfo`, as I proposed in a previous comment, there is no need to decode the whole envelope, just the unprotected header containing the certificate chain.

Reply:
decoding of envelope is required to validate that it contains required signed attributes and there values are valid.